### PR TITLE
CI workflow for Pull Requests

### PR DESCRIPTION
**Pull request by zharinov**
This will enable branch-level tests that are performed under 5 minutes.

Included tests:
- `lein test`
- Uberjar test
- Native binary test (linux-only)

I've selected these tests because they seem comprehensive enough to me, so that you can be 95% confident it's ready for the release.

The entire CI workflow can be quite a big, so let's implement it with the smaller steps. This is the first one.

Ref: #248

### CODE REVIEW

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  build-jar:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3.0.2
+
+      - name: Version
+        id: version
+        # Get version from project.clj
```
Comment (zharinov) [inline .github/workflows/pull-request.yml:15]: There should be more graceful way to obtain current version

Comment (kkinnear) [reply .github/workflows/pull-request.yml:15]: I'm tied up today, let me get back to you tomorrow.  Thanks for keeping going on this!

Comment (kkinnear) [reply .github/workflows/pull-request.yml:15]: Given what I've seen in the example you provided before, I expect that this will work.  I don't have nearly the experience to look at what you've done and have any idea if it is actually correct, since this is the first workflow I've ever examined.  But that is ok.  I would like to accept this PR, but I do have one question.

It looks like it operates whenever someone does a pull request on master?

So, how do you expect this to be used?  I rarely do pull requests on master (or anything).  I expect that is in part because I'm not working with a bunch of other people who add code to the same GitHub repo.   Which could change, of course. 

The workflow you had in #249 was on push or pull with master.  I understand the "push" part, but even there I am kind of not understanding the "pull" trigger.

So I guess I'm asking what is the human workflow you imagine this pull request to be fitting into?  Could you lay out the steps you imagine someone taking to initiate this workflow after I accept this PR?

Once we work this out, I'm ready to accept this PR.

Thanks!

Comment (zharinov) [reply .github/workflows/pull-request.yml:15]: This workflow is for people like me who may want to occasionally contribute to the project by creating Pull Requests. Once all checks are green, both you and me are pretty confident that new code is safe to merge.

This workflow is lightweight enough to be finished under 3–5 minutes, so if the feature is not that trivial, you can push and work further. Distraction will occur only if something went wrong.

For example, GraalVM builds may break because of missing type-hint. While we tend to check our code with unit tests before any code push, errors related to native-image builds are most likely to be discovered during release.

To sum it up, the Pull Request workflow is meant to provide quick enough feedback on the ~95% issues that may happen after particular code change.

To the another hand, workflow for the main branch push should perform more comprehensive set of checks. It will include everything included in Pull Request workflow, but also will run test for ClojureScript version that lasts 7–10 minutes. Once tests of this workflow are done, it should mean we can tag and release this commit.

Finally, release workflow will run `native-image` build on MacOS, Windows and Linux machines, so that you don't need to perform it on your own machines.

Comment (kkinnear) [reply .github/workflows/pull-request.yml:15]: Thanks, that is clearer.  I think my problem was that I was thinking about "push" and "pull" commands to git on my laptop relating to the master branch on GitHub.  I was wondering why I would want a workflow to run when I typed `git pull origin master` on my laptop.  But I don't believe that is what this workflow is for. 

I wasn't thinking about "pull requests", where someone wants to contribute to zprint by doing a PR.  Just to be clear (to me, at least), I think you are saying that this workflow will run when someone creates a PR for zprint?  So, in this case, it is for the person creating the PR to see if the basic tests run, and for me to also see that the PR meets the basic tests? 

Thanks.

Comment (zharinov) [reply .github/workflows/pull-request.yml:15]: You're right, this meant to be the very first criteria for successful Pull Request. This is not necessary should be a human, but also can be Renovate Bot with its dependency upgrades.

One reason why people don't upgrade dependencies often IMO is that we're subconsciously afraid of unexpected work of researching which one new package update breaks our project. And if we don't have any CI, sometimes we risk to learn about breakage from the users opening an issue, which makes us feel guilty so at some point we decided: "I'll never ever will upgrade dependencies unless it's absolutely necessary".

But once this is automated, you'll transparently see which upgrade breaks the project and which is okay to merge.

By the way, I noticed from personal experience that once these PR workflows are implemented, I tend to work using Pull Requests, despite the fact direct pushes are quicker and seemed to be pragmatic for the projects with one contributor.

**Comment by kkinnear** [comment]
So, If I create a PR myself, and the tests fail, what's next?  Do I cancel (drop or whatever) that PR and create a new PR from scratch with the better code, or do I edit the PR and then the tests automatically run again when the PR changes?

**Comment by zharinov** [comment]
No, you just push code fix commits to the same branch and tests will re-run. It's very convenient to use "Squash and merge"

**Comment by kkinnear** [comment]
Thanks for the info, I tried out the workflow with my own PR #260 and it seems to work great!   Note that I also changed the default branch from `master` to `main` while I was at it.  I have updated the workflow so it matches `main`.

**Comment by kkinnear** [comment]
This is really great work!  I used it for the first time (other than the test I did above), and it worked out really well.  I even managed to recover the `zprintl-1.2.5` and bring it local and run some additional testing on it just to be sure, and it worked flawlessly!  I may need to send the intermediate linux `zprintl-1.2.5` to someone to fix a serious bug before I'm ready to release the whole thing.

It seems like if I just leave a PR open from the branch on which I'm working, then every time I push that branch to GitHub I'll gets tests, and when it is the "final" push, I'll accept the PR into main and then pull main back down to my local repo for some final work.

Anyway, this is already starting to change the way I work, and I want to thank you for all of the effort you put into make all of this happen.  It is very impressive, and I'm not sure I would have ever taken the time to figure out how to do it all.  Thank you so much!

**Comment by arichiardi** [comment]
Wanted to add, thanks for doing this!
